### PR TITLE
100 hooktrigger methods

### PR DESCRIPTION
**Pull request by jonsberg**

### CODE REVIEW

File: platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/trigger/CreateTriggerHandlerTest.kt [hunk 1]
```diff
@@ -92,7 +96,37 @@ internal class CreateTriggerHandlerTest : BaseReqHandlerTest() {
         }
 
         @Test
-        fun `Tries to create trigger, but func does not exists`() {
+        fun `Tries to create an invalid trigger`() {
```
Comment (dominique-dymke) [inline platform/backend/core/src/integrationTest/kotlin/io/hamal/core/req/handler/trigger/CreateTriggerHandlerTest.kt:99]: This test name does not describe the problem case you want to test - as in: What is an invalid trigger in this case

File: platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/controller/trigger/TriggerCreateControllerTest.kt [hunk 2]
```diff
@@ -301,9 +305,64 @@ internal class TriggerCreateControllerTest : TriggerBaseControllerTest() {
...
 
+
+        @Test
+        fun `Creates valid triggers`() {
+            val hook_one = awaitCompleted(createHook(HookName("hook-name-one"))).hookId
+            val hook_two = awaitCompleted(createHook(HookName("hook-name-two"))).hookId
+            val func_one = awaitCompleted(createFunc(FuncName("hook-trigger-func-one"))).funcId
+            val func_two = awaitCompleted(createFunc(FuncName("hook-trigger-func-two"))).funcId
+
+
```
Comment (dominique-dymke) [inline platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/controller/trigger/TriggerCreateControllerTest.kt:320]:
```suggestion

```

File: platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/controller/trigger/TriggerCreateControllerTest.kt [hunk 3]
```diff
@@ -301,9 +305,64 @@ internal class TriggerCreateControllerTest : TriggerBaseControllerTest() {
                 assertThat(this.funcId, equalTo(funcId))
                 require(this is HookTrigger) { "not HookTrigger" }
                 assertThat(this.hookId, equalTo(hookId))
+                assertThat(this.hookMethod, equalTo(Post))
             }
         }
 
+
+        @Test
+        fun `Creates valid triggers`() {
```
Comment (dominique-dymke) [inline platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/controller/trigger/TriggerCreateControllerTest.kt:314]: A better test name would be great, which describes the why

File: platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/controller/trigger/TriggerCreateControllerTest.kt [hunk 4]
```diff
@@ -326,6 +385,75 @@ internal class TriggerCreateControllerTest : TriggerBaseControllerTest() {
             verifyNoRequests(TriggerCreateSubmitted::class)
         }
 
+
+        @Test
+        fun `Tries to create invalid trigger`() {
```
Comment (dominique-dymke) [inline platform/backend/api/src/integrationTest/kotlin/io/hamal/api/http/controller/trigger/TriggerCreateControllerTest.kt:390]: Same test name - what makes the trigger invalid

File: platform/backend/repository/sqlite/src/main/kotlin/io/hamal/repository/sqlite/record/trigger/Repository.kt [hunk 5]
```diff
@@ -118,6 +118,19 @@ class SqliteTriggerRepository(
...
+                    TriggerQuery(
+                        hookIds = listOf(cmd.hookId),
+                        funcIds = listOf(cmd.funcId)
+                    )
+                ).firstOrNull()?.let { trigger ->
+                    trigger as HookTrigger
+                    if (trigger.hookMethod == cmd.hookMethod) {
+                        throw IllegalArgumentException("Trigger already exists")
+                    }
+                }
```
Comment (dominique-dymke) [inline platform/backend/repository/sqlite/src/main/kotlin/io/hamal/repository/sqlite/record/trigger/Repository.kt:132]: Good idea - but this should be handled differently:

==> on sqlite level with an partial unique index

see ProjectionUniqueName as reference implementation

Do the same thing for Memory

File: platform/backend/repository/memory/src/main/kotlin/io/hamal/repository/memory/record/Trigger.kt [hunk 6]
```diff
@@ -169,6 +169,19 @@ class MemoryTriggerRepository : MemoryRecordRepository<TriggerId, TriggerRecord,
...
+                        hookIds = listOf(cmd.hookId),
+                        funcIds = listOf(cmd.funcId)
+                    )
+                ).firstOrNull()?.let { trigger ->
+                    trigger as HookTrigger
+                    if (trigger.hookMethod == cmd.hookMethod) {
+                        throw IllegalArgumentException("Trigger already exists")
+                    }
+                }
+
```
Comment (dominique-dymke) [inline platform/backend/repository/memory/src/main/kotlin/io/hamal/repository/memory/record/Trigger.kt:184]: see sqlite implementation comment

File: platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/TriggerTest.kt [hunk 7]
```diff
@@ -429,13 +431,85 @@ internal class TriggerRepositoryTest : AbstractUnitTest() {
...
+                funcId = FuncId(2),
+                hookId = HookId(1),
+                hookMethod = Get
+            )
+
+            verifyCount(4)
+        }
+
+        @TestFactory
+        fun `Tries to register invalid hook triggers`() = runWith(TriggerRepository::class) {
```
Comment (dominique-dymke) [inline platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/TriggerTest.kt:487]: what makes those hook trigger invalid?

File: platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/TriggerTest.kt [hunk 8]
```diff
@@ -429,13 +431,85 @@ internal class TriggerRepositoryTest : AbstractUnitTest() {
...
-                    assertThat(hookMethods, equalTo(setOf(Post, Put)))
+                    assertThat(hookMethod, equalTo(Post))
                     assertThat(status, equalTo(Active))
                 }
 
                 verifyCount(2)
             }
 
+        @TestFactory
+        fun `Registers valid triggers`() = runWith(TriggerRepository::class) {
```
Comment (dominique-dymke) [inline platform/backend/repository/testbed/src/test/kotlin/io/hamal/repository/TriggerTest.kt:442]: same test name

File: platform/runner-plugin/std/sys/src/main/kotlin/io/hamal/plugin/std/sys/topic/trigger/TriggerCreateFunction.kt [hunk 9]
```diff
@@ -44,12 +45,23 @@ class TriggerCreateFunction(
...
                         null
                     },
+                    hookMethod = if (arg1.type("hook_method") == StringType::class) {
+                        when (arg1.getString("hook_method")) {
+                            "DELETE" -> Delete
+                            "GET" -> Get
+                            "PATCH" -> Patch
+                            "POST" -> Post
+                            "PUT" -> Put
+                            else -> null
```
Comment (dominique-dymke) [inline platform/runner-plugin/std/sys/src/main/kotlin/io/hamal/plugin/std/sys/topic/trigger/TriggerCreateFunction.kt:55]:
```
hookMethod = arg1.findString("hook_method")?.let(HookMethod::valueOf),
```

can be simplified as - we want to use the same Name in lua as in kotlin

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/create/create_invalid/01.lua [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+hook_req = fail_on_error(sys.hooks.create({ name = "some-amazing-hook" }))
+sys.await_completed(hook_req)
+
+trigger_req = fail_on_error(sys.triggers.create_hook({
+    func_id = func_req.func_id,
+    flow_id = '1',
+    name = 'trigger-one',
+    inputs = { },
+    hook_id = hook_req.hook_id,
+    hook_method = 'GET'
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/create/create_invalid/01.lua:15]:
```suggestion
    hook_method = 'Get'
```

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/create/create_invalid/01.lua [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+}))
+sys.await_completed(trigger_req)
+
+err, trigger = sys.triggers.create_hook({
+    func_id = func_req.func_id,
+    flow_id = '1',
+    name = 'trigger-two',
+    inputs = { },
+    hook_id = hook_req.hook_id,
+    hook_method = 'GET'
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/create/create_invalid/01.lua:25]:
```suggestion
    hook_method = 'Get'
```

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/create/current_flow/01.lua [hunk 12]
```diff
@@ -10,6 +10,7 @@ trigger_req = fail_on_error(sys.triggers.create_hook({
     func_id = func_req.func_id,
     name = 'hook-trigger',
     inputs = { },
+    hook_method = 'GET',
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/create/current_flow/01.lua:13]:
```suggestion
    hook_method = 'Get',
```

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/create/different_flow/01.lua [hunk 13]
```diff
@@ -14,6 +14,7 @@ trigger_req = fail_on_error(sys.triggers.create_hook({
     func_id = func_req.id,
     name = 'hook-trigger',
     inputs = { },
+    hook_method = 'GET',
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/create/different_flow/01.lua:17]:
```suggestion
    hook_method = 'Get',
```

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/list/01.lua [hunk 14]
```diff
@@ -10,6 +10,7 @@ trigger_req = fail_on_error(sys.triggers.create_hook({
     func_id = func_req.func_id,
     name = 'trigger-to-create',
     inputs = { },
+    hook_method = 'GET',
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/list/01.lua:13]:
```suggestion
    hook_method = 'Get',
```

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/name_unique_in_flow/01.lua [hunk 15]
```diff
@@ -15,7 +15,9 @@ trigger = fail_on_error(sys.triggers.create_hook({
     flow_id = '1',
     name = 'trigger-to-create',
     inputs = { },
-    hook_id = hook.id
+    hook_id = hook.id,
+    hook_method = 'GET'
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/name_unique_in_flow/01.lua:19]:
```suggestion
    hook_method = 'Get'
```

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/name_unique_in_flow/01.lua [hunk 16]
```diff
@@ -24,7 +26,8 @@ trigger = fail_on_error(sys.triggers.create_hook({
     flow_id = '1',
     name = 'trigger-to-create',
     inputs = { },
-    hook_id = hook.id
+    hook_id = hook.id,
+    hook_method = 'POST'
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/name_unique_in_flow/01.lua:30]:
```suggestion
    hook_method = 'Post'
```

File: platform/testbed/src/integrationTest/resources/api/trigger/hook/name_unique_in_flow/01.lua [hunk 17]
```diff
@@ -37,10 +40,11 @@ err, trigger = sys.triggers.create_hook({
     flow_id = flow.id,
     name = 'trigger-to-create',
     inputs = { },
-    hook_id = hook.id
+    hook_id = hook.id,
+    hook_method = 'PATCH'
```
Comment (dominique-dymke) [inline platform/testbed/src/integrationTest/resources/api/trigger/hook/name_unique_in_flow/01.lua:44]:
```suggestion
    hook_method = 'Patch'
```

**Review by dominique-dymke — CHANGES_REQUESTED** [review]

**Comment by jonsberg** [comment]
- SQLite implementation --> ProjectionUniqueHook (separate table, unique constraints), so ProjectionCurrent is no longer bothered with this
- memory implementation --> MutableSet
- introduced a method ensureHookUnique to TriggerQueryRepo, that utilizes this Table/Set to ensure the client receives a BadRequest before the operation 
- if this is bs I'll remove it again

**Review by dominique-dymke — APPROVED** [review]
well done

**Pull request merged**